Add processing model to spec (#270)
w3c/distributed-tracing: spec/30-processing-model.md

> # Processing Model

Reviewer: please use some plugin for fixed length lines. It would be easier to comment

Reviewer 2: [EditorConfig](https://editorconfig.org/) is considering adding [support for this](https://github.com/editorconfig/editorconfig/wiki/Property-research:-maximum-line-length). We may want to act sooner, but EditorConfig files would be the most democratic way to solve this issue as it's editor agnostic.

> This section provides a step-by-step description of the behavior of a tracing implementation - also called tracer -  when a request is received. This description can be used as a reference when implementing a tracecontext compliant tracing system, middleware or cloud service.

Reviewer: tracecontext should be written as two words

Reviewer: ```suggestion
This section provides a step-by-step description of the behavior of a tracing implementation - also called tracer -  when a request is received. This description can be used as a reference when implementing a Trace Context compliant tracing system, middleware or cloud service.
```

> This section provides a step-by-step description of the behavior of a tracing implementation - also called tracer -  when a request is received. This description can be used as a reference when implementing a tracecontext compliant tracing system, middleware or cloud service.

Reviewer: >when a request is received

perhaps add a bit more clarification. There is a mention of outgoing request in processing model. Perhaps say something like: "This processing model assumes that there are one or more outgoing requests would happen during the processing of incoming request."

Author: How would this help?

> This section provides a step-by-step description of the behavior of a tracing implementation - also called tracer -  when a request is received. This description can be used as a reference when implementing a tracecontext compliant tracing system, middleware or cloud service.

Reviewer: one thing confused me - beginning saying that this is tracing implementation behavior. And later you recommend to use it for middleware. Perhaps we understand middleware differently, can you please elaborate what you meant by middleware.

Author: middleware as in proxy - added examples in latest commit

## Processing Model for Working with Trace Context

> The processing model describes the behaviour of a tracer which forwards **and** modifies

Reviewer: why do you need bold for "and"?

Author: don't need was there in the previous version. is gone now

> The processing model describes the behaviour of a tracer which forwards **and** modifies

Reviewer: ```suggestion
The processing model describes the behavior of a tracer which forwards **and** modifies
```

> The processing model describes the behaviour of a tracer which forwards **and** modifies

Reviewer 2: We should decide if we want to use the American or British spelling of words throughout the document. This is a nit to be sure, but something to consider for a consistent tone. We used the American spelling the line above, but the British one here. Of course I'm biased, so I'd choose the American spelling - because 'MERICA! 😜  USA! USA! USA! 

```suggestion
The processing model describes the behavior of a tracer which forwards **and** modifies
```

Author: America is fine.  I trust you to nativy the language :-)

Author: I cannot merge this one as it is outdated now, but agree on US English

> tracecontext headers.

Reviewer: ```suggestion
Trace Context headers.
```

> tracecontext headers.

Reviewer: should we use fields instead of headers? So this section will be more like a semantical model that includes other protocols?

Author: the main document is still about HTTP so I think headers is fine b/c that's what it is. Alternative protocols should define how they translate headers to other transport mechanisms

Reviewer: it's fine to keep headers. Maybe later down the release cycle we can address this. Now we have them as fields in spec in some places and headers in others

> 1. The tracer receives a `traceparent` header and may receive a `tracestate`. header.

Reviewer: dot before word header is not required

Reviewer: ```suggestion
1. The tracer receives a `traceparent` header and may receive a `tracestate` header.
```

> 1. The tracer receives a `traceparent` header and may receive a `tracestate` header.

Reviewer 2: Isn't it true that the tracer *may* receive a `traceparent` as well? I believe this entire model is predicated on receiving that header, and if it's absent the tracer would effectively exit the routine (or potentially start a new trace).

```suggestion
1. The tracer may receive a `traceparent` header and/or may receive a `tracestate` header.
```

Author: Valid point. I think I will add a section for starting a trace separately. This still adds the question on what to do when receiving a trace-state without a trace-parent. It can be that there was a proxy or the request is broken.

Reviewer 2: Right. My understanding it that `traceparent` is required. An orphaned `tracestate` could/should be thrown away - just like if the `tracestate` is accompanied with an invalid `traceparent`.

Author: What do you think about the current rewrite.

> 2. The tracer tries to parse the version of the `traceparent` header.

Reviewer 2: ```suggestion
2. If received, the tracer tries to parse the version of the `traceparent` header.
```

Reviewer 2: If we accept this suggested change, we may want to say something to the effect of "If `traceparent` is not received, a trace may be started".

Author: I think this falls under the bigger topic of making starting a trace part of the process model - right?

Reviewer 2: Correct

Author: this should be handled in the latest version

   - If the version prefix cannot be parsed, the tracer creates a new `traceparent`
     header and removes all entries from `tracestate`.
   - If the version number is higher than supported by the tracer,

>   the implementation uses the format defined in this specification to parse

Reviewer: this should have four spaces to indicate it's a sublist continuation

>   `trace-id` and `parent-id`. The tracer will only parse `flags` values

Reviewer: Use ` Trace-flags` for consistency

Reviewer: ```suggestion
  `trace-id` and `parent-id`. The tracer will only parse `trace-flags` values
```

Author: I copied this from your text, but can change

  supported by the current version of this specification and ignore all other
  values. If parsing fails, the tracing system creates a new `traceparent` header
  and removes all entries from `tracestate`.

> 3. If the tracer supports the version number it validates `trace-id`

Reviewer: ```suggestion
3. If the tracer supports the version number, it validates `trace-id`
```

> and `parent-id`.

Reviewer: should be 2 spaces before and so markdown will be valid

Reviewer: ```suggestion
  and `parent-id`.
```

>    - If either `trace-id`, `parent-id` or `flags`  are invalid,  the tracer

Reviewer 2: Should this be `trace-flags` as well?

```suggestion
   - If either `trace-id`, `parent-id` or `trace-flags`  are invalid,  the tracer
```

   creates a new `traceparent` header and removes all entries from `tracestate`.

> 4. The tracer validates the `tracestate` header. If the `tracestate` header cannot be parsed the system creates a new empty `tracestate` header.

Reviewer 2: I believe this should be a *may* statement. Thoughts?

```suggestion
4. The tracer MAY validate the `tracestate` header. If the `tracestate` header cannot be parsed the system creates a new empty `tracestate` header.
```

Author: If you modify the trace-state you will have to validate. If you leave it alone you would not have to.

> 4. The tracer validates the `tracestate` header. If the `tracestate` header cannot be parsed the system creates a new empty `tracestate` header.

Reviewer 2: My understanding is that the absence of the `tracestate` header is functionally equivalent of a `tracestate` header with an empty value. If my understanding is correct, I'd prefer to not recommend creating empty `tracestate`'s. We can save precious bytes on the wire this way. Thoughts from the group?

Author: works for me

Author: I reworded all text reflecting this.

5. For each outgoing request the tracer performs the following steps:

>    - The tracing system MUST modify the `traceparent` header.

Reviewer 2: I believe the list of items beneath this item (Update `parent-id`, Update `recorded`, ..., etc) would benefit from being an ordered list, if we can get the markdown to comply.

Author: Order lists will switch to using lower case i on the second level. Maybe @plehegar know how to make nested ordered lists work.

>         - **Update `parent-id`**. The value of property `parent-id` MUST be set to        the a value representing the ID of the current operation.

Reviewer: ```suggestion
        - **Update `parent-id`**. The value of property `parent-id` MUST be set to a value representing the ID of the current operation.
```

>         -  **Update `recorded`**. The value of `recorded` reflects the callers

Reviewer 2: This usage is possessive, not plural, necessitating a `'`.

```suggestion
        -  **Update `recorded`**. The value of `recorded` reflects the caller's
```

            recording behavior. The value of the `recorded` flag of `trace-flags`

>             MAY be set to `1` if the trace data is likely to be recorded or to `0` otherwise. Setting the flag is no guarantee that the trace will be recorded but increases the likeliness of end-to-end recorded traces.

Reviewer: Should recorded flag be changed by incoming request "processor"?

Author: you mean move it to outbound calls section?

>             MAY be set to `1` if the trace data is likely to be recorded or to `0` otherwise. Setting the flag is no guarantee that the trace will be recorded but increases the likeliness of end-to-end recorded traces.

Reviewer: Perhaps it may be clearer whether one recording trace data for incoming request or for outgoing

>             MAY be set to `1` if the trace data is likely to be recorded or to `0` otherwise. Setting the flag is no guarantee that the trace will be recorded but increases the likeliness of end-to-end recorded traces.

Reviewer 2: Should we also mention that if the value is `1`, that the tracer should consider logging this trace to the back end? I'm not sure the exact right way to say this.

Author: prioritize might be the word to use maybe?

>     - The tracer modifies the `tracestate` header

Reviewer: Fix spacing:

![image](https://user-images.githubusercontent.com/9950081/55443875-2d36bd80-5569-11e9-9fe8-3e60246d6a07.png)

Reviewer 2: Love the extra mile of providing a screenshot here @SergeyKanzhelev.

Author: This should be fixed, at least to VSCode

Author: should be fixed

        - **Update key value**. The value of any key can be updated. Modified keys
        MUST be moved to the beginning of the list. 
        - **Add new key-value pair**. The new key-value pair should be added into
        the beginning of the list.
        - **Delete the key-value pair**. Any key-value pair MAY be deleted. It is
        highly discouraged to delete keys that weren't generated by the same tracing
        system or platform. Deletion of an unknown key-value pair will break

>         correlation in other systems. This mutation enables two scenarios:

Reviewer: "This mutation enables two scenarios" -> "There are cases when it is needed to delete unknown key-value pairs:". Otherwise it is written like there is no valid scenario to delete your own keys

Reviewer 2: I find this slightly confusing as well. I also question if this statement of reason belongs in the processing model. Perhaps the information about "why" should be in the definition of `tracestate` above?

            - Proxies can block specific `tracestate` keys for privacy and security
            concerns.
            - Truncation of long `tracestate` entries.

>     - The tracer set the `traceparent` and `tracestate` header for the outgoing request

Reviewer: fix spacing:

![image](https://user-images.githubusercontent.com/9950081/55443683-7f2b1380-5568-11e9-92b1-35c9e01e81c8.png)

## Alternative Processing

> The processing model above described the complete set of steps for processing tracecontext headers. There are, however, situation when an implementation only supports a subset of the steps described above. Proxies or messaging middleware MAY not modify the `traceparent` headers but may remove invalid headers or add additional information to `tracestate`. 

Reviewer: "MAY not" sounds bad. It may be read as "MAY NOT" which changes meaning.

Author: decide to not

> The processing model above described the complete set of steps for processing tracecontext headers. There are, however, situation when an implementation only supports a subset of the steps described above. Proxies or messaging middleware MAY not modify the `traceparent` headers but may remove invalid headers or add additional information to `tracestate`. 

Reviewer 2: ```suggestion
The processing model above describes the complete set of steps for processing Trace Context headers. There are, however, situations when an implementation might only support a subset of the steps described above. Proxies or messaging middleware MAY not modify the `traceparent` headers but may remove invalid headers or add additional information to `tracestate`. 
```

# Processing Model

This section provides a step-by-step exemplary description of the behavior of a tracing

> implementation - also called tracer - when a request is received, processed and 

Reviewer 2: ```suggestion
implementation - also called a tracer - illustrating when a request is received, processed and 
```

   and `parent-id` representing the current request.
3. If `traceparent` header is present, the tracer tries to parse the version of
   the `traceparent` header.
    - If the version cannot be parsed, the tracer creates a new
      `traceparent` header and deletes `tracestate`.
    - If the version number is higher than supported by the tracer, the
      implementation uses the format defined in this specification to parse
      `trace-id` and `parent-id`. The tracer will only parse `trace-flags` values
      supported by the current version of this specification and ignore all other
      values. If parsing fails, the tracing system creates a new `traceparent`
      header and deletes the `tracestate`.
4. If the tracer supports the version number, it validates `trace-id` and
   `parent-id`.
       - If either `trace-id`, `parent-id` or `trace-flags` are invalid, the tracer
      creates a new `traceparent` header and deletes `tracestate`.
5. The tracer validates the `tracestate` header. If the `tracestate` header
   cannot be parsed the tracer deletes it.
6. For each outgoing request the tracer performs the following steps:

>     - The tracing system MUST modify the `traceparent` header.

Reviewer 2: ```suggestion
    - The tracing system MUST modify the `traceparent` header in the following ways:
```

        - **Update `parent-id`**. The value of property `parent-id` MUST be set to a
          value representing the ID of the current operation.
        - **Update `recorded`**. The value of `recorded` reflects the caller's
          recording behavior. The value of the `recorded` flag of `trace-flags` MAY
          be set to `1` if the trace data is likely to be recorded or to `0`
          otherwise. Setting the flag is no guarantee that the trace will be
          recorded but increases the likeliness of end-to-end recorded traces.

>     - The tracer MAY modify the `tracestate` header

Reviewer 2: ```suggestion
    - The tracer MAY modify the `tracestate` header in any of the following ways: 
```

forwards Trace Context headers.

1. The tracer checks an incoming request for a `traceparent` and a `tracestate`
   header.
2. If no `traceparent` header is received, the tracer creates a new `trace-id`
   and `parent-id` representing the current request.
3. If `traceparent` header is present, the tracer tries to parse the version of
   the `traceparent` header.
   - If the version cannot be parsed, the tracer creates a new
     `traceparent` header and deletes `tracestate`.
   - If the version number is higher than supported by the tracer, the
     implementation uses the format defined in this specification to parse
     `trace-id` and `parent-id`. The tracer will only parse `trace-flags` values
     supported by the current version of this specification and ignore all other
     values. If parsing fails, the tracing system creates a new `traceparent`
     header and deletes the `tracestate`.
4. If the tracer supports the version number, it validates `trace-id` and
   `parent-id`.
   - If either `trace-id`, `parent-id` or `trace-flags` are invalid, the tracer
     creates a new `traceparent` header and deletes `tracestate`.

> 5. The tracer validates the `tracestate` header. If the `tracestate` header

Reviewer 3: Should we also add something about validating the size of tracestate?

Reviewer 3: Looks good to me otherwise

Author: We decided to leave size handling open and have everyone decide how much they want to forward. So not entirely sure what to add.

Reviewer 3: Makes sense, thanks!

   cannot be parsed the tracer deletes it.
6. For each outgoing request the tracer performs the following steps:
  - The tracing system MUST modify the `traceparent` header.
    - **Update `parent-id`**. The value of property `parent-id` MUST be set to a
       value representing the ID of the current operation.
    - **Update `recorded`**. The value of `recorded` reflects the caller's
      recording behavior. The value of the `recorded` flag of `trace-flags` MAY
      be set to `1` if the trace data is likely to be recorded or to `0`
      otherwise. Setting the flag is no guarantee that the trace will be
      recorded but increases the likeliness of end-to-end recorded traces.
  - The tracer MAY modify the `tracestate` header
     - **Update key value**. The value of any key can be updated. Modified keys
        MUST be moved to the beginning of the list.

>      - **Add new key-value pair**. The new key-value pair should be added into

Reviewer: ```suggestion
     - **Add new key-value pair**. The new key-value pair MUST be added into
```

       the beginning of the list.
     - **Delete the key-value pair**. Any key-value pair MAY be deleted. It is
       highly discouraged to delete keys that weren't generated by the same
       tracing system or platform. Deletion of an unknown key-value pair will
       break correlation in other systems. This mutation enables two scenarios:

>        - Proxies can block specific `tracestate` keys for privacy and security

Reviewer: list broken now. Also, again - deleting of key-value pair has more applications than those two scenarios. If you want to explain that these two scenarios can justify the deletion of unknown keys - please tell this explicitly. Otherwise - let's delete those scenarios from here.

Author: This was your initial text. I can just delete the examples.

Reviewer: Those are valid scenarios for the situation when you are deleting somebody else's tracestate pairs. This whole item about deleting in general, which has more valid use cases. My suggestion is to make it clear that this list of scenarios talking about exceptional cases, not the standard ones

# Processing Model

This section provides a step-by-step description of the behavior of a tracing

> implementation - also called tracer -  when a request is received. This

Reviewer: not sure where my comment went. Here I had a comment:

>when a request is received

Perhaps say that it's a behavior when request is recieved, processed, and optionally outgoing requests initiated during request processing. I don't think it's a safe assumption to expect that this is clear from simple "request is received".

Author: changed.

description can be used as a reference when implementing a Trace Context
compliant tracing system, middleware - like a proxy or messaging bus -  or cloud
service.

## Processing Model for Working with Trace Context

> The processing model describes the behaviour of a tracer which forwards and modifies tracecontext headers.

Reviewer: > which forwards and modifies

ordering is strange. Forwards first, than modifyes?

Author: changed

> The processing model describes the behaviour of a tracer which forwards and modifies tracecontext headers.

Reviewer: ```suggestion
The processing model describes the behaviour of a tracer which forwards and modifies Trace Context headers.
```

1. The tracer checks an incoming request for a `traceparent` and a `tracestate`
   header.

> 2. If no `traceparent` header is received, the trace creates a new `trace-id`

Reviewer: trace is not something that can create. Trace being created?

Reviewer: oh, perhaps you meant tracer here. 

```suggestion
2. If no `traceparent` header is received, the tracer creates a new `trace-id`
```

>    and `parent-id` representing the current opeartion.

Reviewer 4: Minor typo, `opeartion` => `operation`

Reviewer 5: Fixed

>    and `parent-id` representing the current operation.

Reviewer: ```suggestion
   and `parent-id` representing the current request.
```

Reviewer: perhaps request is a better term here? Term operation was not defined before and it may be ambiguous

> 1. If `traceparent` header is present, the tracer tries to parse the version of the `traceparent` header.

Reviewer: should be 3?

```suggestion
3. If `traceparent` header is present, the tracer tries to parse the version of the `traceparent` header.
```

>    - If the version prefix cannot be parsed, the tracer creates   a new `traceparent` header and deletes `tracestate`.

Reviewer: Perhaps don't say "prefix". It may be read as `version` field has a prefix.

```suggestion
   - If the version cannot be parsed, the tracer creates   a new `traceparent` header and deletes `tracestate`.
```

   - If the version number is higher than supported by the tracer, the
     implementation uses the format defined in this specification to parse
     `trace-id` and `parent-id`. The tracer will only parse `trace-flags` values
     supported by the current version of this specification and ignore all other
     values. If parsing fails, the tracing system creates a new `traceparent`
     header and deletes the `tracestate`.

3. If the tracer supports the version number, it validates `trace-id` and
   `parent-id`.

>    - If either `trace-id`, `parent-id` or `trace-flags`  are invalid,  the tracer

Reviewer: why to have a list here at all? Just continue the paragraph

Author: readability

     creates a new `traceparent` header and deletes `tracestate`.

4. The tracer validates the `tracestate` header. If the `tracestate` header
   cannot be parsed the tracer deletes it.

5. For each outgoing request the tracer performs the following steps:

   - The tracing system MUST modify the `traceparent` header.
        - **Update `parent-id`**. The value of property `parent-id` MUST be set
          to a value representing the ID of the current operation.
        -  **Update `recorded`**. The value of `recorded` reflects the caller's
            recording behavior. The value of the `recorded` flag of
            `trace-flags` MAY be set to `1` if the trace data is likely to be
            recorded or to `0` otherwise. Setting the flag is no guarantee that
            the trace will be recorded but increases the likeliness of
            end-to-end recorded traces.

>     - The tracer MAY modify the `tracestate` header

Reviewer: spacing is broken here:

![image](https://user-images.githubusercontent.com/9950081/55831067-23f8a400-5ac7-11e9-9be0-103488835187.png)

        - **Update key value**. The value of any key can be updated. Modified
          keys MUST be moved to the beginning of the list.
        - **Add new key-value pair**. The new key-value pair should be added
          into the beginning of the list.
        - **Delete the key-value pair**. Any key-value pair MAY be deleted. It
          is highly discouraged to delete keys that weren't generated by the
          same tracing system or platform. Deletion of an unknown key-value pair

>           will break correlation in other systems. This mutation enables two

Reviewer: again, comment here were lost. This mutation enabled more than those two scenarios. These twoi scenarios only clarifies why anybody may want to change entries they don't own

Author: this was taken from your original text.

          scenarios:
            - Proxies can block specific `tracestate` keys for privacy and
              security concerns.
            - Truncation of long `tracestate` entries.

>     - The tracer set the `traceparent` and `tracestate` header for the outgoing

Reviewer 6: I think there is a word missing here. Should it be "The tracer sets", or possibly "The tracer SHOULD/MUST set"?

Author: given what we have in the privacy statement it would be MAY set but sets should be generally fine

Author: changed

>     - The tracer set the `traceparent` and `tracestate` header for the outgoing

Reviewer: Again, spacing is broken:

![image](https://user-images.githubusercontent.com/9950081/55831195-64582200-5ac7-11e9-83af-41eea5637a5b.png)

Author: This should be resolved now

>     - The tracer set the `traceparent` and `tracestate` header for the outgoing

Reviewer: Perhaps entire line: `The tracer set the `traceparent` and `tracestate` header for the outgoing` can be moved to the first item before all the possible mutations list. Something like:

```
5. For each outgoing request the tracer needs to set Trace Context headers. Value of these headers should be generated from the values recieved for incoming requests following these rules: 
```

This will put an action item in the beggiing so details on how mutations should happen would be easier to navigate

      request

## Alternative Processing

The processing model above describes the complete set of steps for processing
Trace Context headers. There are, however, situations when an implementation
might only support a subset of the steps described above. Proxies or messaging

> middleware MAY decide to not modify the `traceparent` headers but may remove invalid

Reviewer: uppercased "MAY"? Or just no "may" at all as it is an example that already started with MAY.

```suggestion
middleware MAY decide to not modify the `traceparent` headers but remove invalid
```
